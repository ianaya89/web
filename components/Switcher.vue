<template>
  <div class="switcher">
    <v-switch :input-value="value" v-bind="$props" :disabled="disabled" @change="onChange" />
  </div>
</template>

<script>
import BaseInput from '@/components/BaseInput'

export default BaseInput.extend({
  name: 'Switcher',

  props: {
    color: {
      type: String,
      default: 'red'
    },
    value: {
      type: Boolean
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },

  methods: {
    onChange(value) {
      this.$emit('input', value || false)
    }
  }
})
</script>

<style lang="postcss">
.switcher {
  mb-6 sm:mb-0 .v-input--selection-controls {
    @apply mt-0;
  }

  .v-input__control {
    min-height: auto;
    .v-input--selection-controls__input {
      @apply .mr-10;
    }

    .v-label {
      @apply .mt-1 .font-serif .normal-case;
    }

    .v-input--switch__thumb {
      @apply .w-6 .h-6 .shadow-none .bg-grey-l2;
      left: 7px;

      &.red--text {
        @apply .bg-red;
      }

      &.green--text {
        @apply .bg-green;
      }
    }

    .v-input--switch__track {
      @apply .bg-grey-l3 .h-8;
      width: 3.5rem;
      border-radius: 1.5rem;
      top: calc(50% - 0.875rem);

      &.red--text {
        @apply .bg-red-l5;
      }

      &.green--text {
        @apply .bg-green-l4;
      }
    }
    .v-input--is-disabled {
      .v-input--selection-controls__input {
        @apply .opacity-25;
      }
    }

    .v-input--selection-controls__input {
      transition: 0.2s cubic-bezier(0.25, 0.8, 0.25, 1);
    }
  }

  .v-input--switch .v-input--selection-controls__ripple {
    left: -2px;
    opacity: 0.4;
    top: calc(50% - 1.375rem);
  }

  .v-input--switch.v-input--is-dirty {
    .v-input--switch__thumb {
      transform: translate(1.375rem, 0);
    }
  }
}

@media (max-width: 768px) {
  .switcher {
    .v-input--selection-controls {
      @apply .mt-3;
    }
  }
}
</style>
