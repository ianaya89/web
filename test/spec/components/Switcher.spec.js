import { mount, shallowMount } from '@vue/test-utils'
import Switcher from '@/components/Switcher'

describe('Switcher', () => {
  it('renders default switcher', () => {
    const wrapper = shallowMount(Switcher)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('switcher test value', () => {
    const wrapper = mount(Switcher, {
      propsData: {
        value: true
      }
    })

    expect(wrapper.element).toMatchSnapshot()
    wrapper.find('input').setChecked(false)
    expect(wrapper.emitted('input')).toBeTruthy()
    expect(wrapper.emitted('input')[0][0]).toBeFalsy()
  })

  it('disabled switcher', () => {
    const wrapper = shallowMount(Switcher, {
      propsData: {
        disabled: true,
        label: 'My label'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
