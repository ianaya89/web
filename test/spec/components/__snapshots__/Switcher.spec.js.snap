// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher disabled switcher 1`] = `
<div
  class="switcher"
>
  <v-switch-stub
    backgroundcolor=""
    color="red"
    disabled="true"
    errorcount="1"
    errormessages=""
    label="My label"
    messages=""
    ripple="true"
    rules=""
    successmessages=""
    valuecomparator="function deepEqual(a, b) {
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) {
        // If the values are Date, they were convert to timestamp with getTime and compare it
        if (a.getTime() !== b.getTime()) return false;
    }
    if (a !== Object(a) || b !== Object(b)) {
        // If the values aren't objects, they were already checked for equality
        return false;
    }
    var props = Object.keys(a);
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false;
    }
    return props.every(function (p) {
        return deepEqual(a[p], b[p]);
    });
}"
  />
</div>
`;

exports[`Switcher renders default switcher 1`] = `
<div
  class="switcher"
>
  <v-switch-stub
    backgroundcolor=""
    color="red"
    errorcount="1"
    errormessages=""
    label="Label"
    messages=""
    ripple="true"
    rules=""
    successmessages=""
    valuecomparator="function deepEqual(a, b) {
    if (a === b) return true;
    if (a instanceof Date && b instanceof Date) {
        // If the values are Date, they were convert to timestamp with getTime and compare it
        if (a.getTime() !== b.getTime()) return false;
    }
    if (a !== Object(a) || b !== Object(b)) {
        // If the values aren't objects, they were already checked for equality
        return false;
    }
    var props = Object.keys(a);
    if (props.length !== Object.keys(b).length) {
        // Different number of props, don't bother to check
        return false;
    }
    return props.every(function (p) {
        return deepEqual(a[p], b[p]);
    });
}"
  />
</div>
`;

exports[`Switcher switcher test value 1`] = `
<div
  class="switcher"
>
  <div
    class="v-input v-input--selection-controls v-input--switch v-input--is-label-active v-input--is-dirty theme--light"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <div
          class="v-input--selection-controls__input"
        >
          <input
            aria-checked="true"
            aria-label="Label"
            role="checkbox"
            type="checkbox"
            value="true"
          />
          <div
            class="v-input--selection-controls__ripple red--text"
          />
          <div
            class="v-input--switch__track theme--light red--text"
          />
          <div
            class="v-input--switch__thumb theme--light red--text"
          >
            <transition-stub
              mode="out-in"
              name="fab-transition"
            />
          </div>
        </div>
        <label
          aria-hidden="true"
          class="v-label theme--light"
          style="left: 0px; position: relative;"
        >
          Label
        </label>
      </div>
      <div
        class="v-messages theme--light"
      >
        <transition-group-stub
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;
